src/rate_limiter.rs

```diff
@@ -65,8 +66,8 @@ impl RateLimiter {
                        -> Result<(), RoutingError> {
         self.update();
         let total_used: u64 = self.used.values().sum();
-        let used = self.used.entry(client_ip).or_insert(0);
-        let allowance = cmp::min(CAPACITY - total_used, CAPACITY / online_clients - *used);
+        let used = self.used.get(&client_ip).map_or(0, |used| *used);
```

Comment:
Good point - but this now means we can change to `client_ip: &IpAddr` in the function signature.